routing/src/interface.rs

```diff
@@ -0,0 +1,363 @@
+    /// Add an interface to the table
+    //////////////////////////////////////////////////////////////////
+    pub fn add_interface(&mut self, iface: Interface) {
+        self.0.insert(iface.ifindex, iface);
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Remove an interface from the table
+    //////////////////////////////////////////////////////////////////
+    pub fn del_interface(&mut self, ifindex: u32) {
+        self.0.remove(&ifindex);
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Get interface entry from IfTable
+    //////////////////////////////////////////////////////////////////
+    pub fn get_interface(&self, ifindex: u32) -> Option<&Interface> {
+        self.0.get(&ifindex)
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Get interface entry from IfTable, mutably
+    //////////////////////////////////////////////////////////////////
+    pub fn get_interface_mut(&mut self, ifindex: u32) -> Option<&mut Interface> {
+        self.0.get_mut(&ifindex)
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Assign an Ip address to an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn add_ifaddr(&mut self, ifindex: IfIndex, ifaddr: &IfAddress) -> Result<(), RouterError> {
+        if let Some(iface) = self.0.get_mut(&ifindex) {
+            iface.add_ifaddr(ifaddr);
+            Ok(())
+        } else {
+            Err(RouterError::NoSuchInterface(ifindex))
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Un-assign an Ip address from an interface.
+    //////////////////////////////////////////////////////////////////
+    pub fn del_ifaddr(&mut self, ifindex: IfIndex, ifaddr: &IfAddress) {
+        if let Some(iface) = self.0.get_mut(&ifindex) {
+            iface.del_ifaddr(&(ifaddr.0, ifaddr.1));
+        }
+        // if interface does not exist or the address was not configured,
+        // we'll do nothing
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Detach all interfaces attached to some VRF
+    //////////////////////////////////////////////////////////////////
+    pub fn detach_vrf_interfaces(&mut self, vrfid: VrfId) {
+        self.0
+            .values_mut()
+            .filter(|iface| iface.get_vrf().is_some_and(|vrf| vrf.vrfid == vrfid))
+            .for_each(|iface| iface.detach());
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::{IfDataDot1q, IfDataEthernet, IfState, IfTable, IfType, Interface, Vrf};
+    use net::eth::mac::Mac;
+    use net::vlan::Vid;
+    use std::net::IpAddr;
+    use std::str::FromStr;
+    use std::sync::Arc;
+
+    // create a test interface table
+    pub fn build_test_iftable() -> IfTable {
+        /* create interface table */
```

Comment:
This is a good start, but ideally we would implement `TypeGenerator` or `ValueGenerator` of types like `IfTable` and then test properties more abstractly.